File: bookstore/mern/services/publisher.service.js
import PublisherModel from '../models/publisher.model.js'

export default class PublisherService {
  async get() {
    return PublisherModel.find({}).lean()
  }

  async getOne(_id) {
    return PublisherModel.findById(_id).lean()
  }

  async create(name, address) {
    const publisher = await new PublisherModel({ name, address })
    await publisher.save()
    return publisher
  }

  async change(_id, name, address) {
    const publisher = await PublisherModel.findByIdAndUpdate(_id, {
      $set: { name, address },
    }).lean()
    return { ...publisher, name, address }
  }

  async delete(_id) {
    await PublisherModel.findByIdAndRemove(_id)
    return _id
  }
}

File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/services/PublisherService.kt
package masha.bookstore.services

import masha.bookstore.core.BaseService
import masha.bookstore.models.*
import masha.bookstore.interfaces.IPublisherService
import masha.bookstore.repositories.PublisherRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class PublisherService(
    @Autowired override val repository: PublisherRepository
) : BaseService<Publisher, PublisherRepository>(repository), IPublisherService {

    override fun create(name: String, address: String): Publisher {
        val publisher = Publisher(name, address)
        repository.save(publisher)
        return publisher
    }

    override fun change(id: Int, name: String, address: String): Publisher {
        val publisher: Publisher = getOne(id)
        publisher.name = name
        publisher.address = address
        repository.save(publisher)
        return publisher
    }

    override fun addBook(book: Book) {
        book.publisher.let {
            it.books.add(book)
            repository.save(it)
        }
    }

    override fun removeBook(book: Book) {
        book.publisher.let {
            it.books.remove(book)
            repository.save(it)
        }
    }
}